File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result= new ArrayList<>();
        int n=nums.length;
        List<Integer>vis =new ArrayList<>(Collections.nCopies(n,0));
        getAllPermutation(0,n,nums,vis,new ArrayList<>(),result);
        return result;

    }

    private void getAllPermutation(int idx, int n, int[] nums, List<Integer> vis, List<Integer>temp, List<List<Integer>> result) {
        if(idx==n){
            result.add(new ArrayList<>(temp));
            return;
        }
        for(int i=0;i<n;i++){
            if(vis.get(i)==0){
                vis.set(i,1);
                temp.add(nums[i]);
                getAllPermutation(idx+1,n,nums,vis,temp,result);
                vis.set(i,0);
                temp.remove(temp.size()-1);
            }
        }
    }
}

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>>result;
        int n=nums.size();
        vector<int>vis(n,0);
        vector<int>temp;
        sort(nums.begin(),nums.end());
        getAllPermutation(0,n,nums,vis,temp,result);
        return result;        
    }
    
    void getAllPermutation(int idx,int n,vector<int>&nums,vector<int>vis,vector<int>temp,vector<vector<int>>&result){
        if(idx==n){
            result.push_back(temp);
            return;
        }
        for(int i=0;i<n;i++){
            if(!vis[i]){
                vis[i]=1;
                temp.push_back(nums[i]);
                getAllPermutation(idx+1,n,nums,vis,temp,result);
                temp.pop_back();
                vis[i]=0;
            }
            
        }
    }
};